File: source/class/cv/ui/manager/model/Schema.js
/* Schema.js
 *
 * copyright (c) 2010-2022, Christian Mayer and the CometVisu contributers.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 3 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
 */

/**
 *
 */
qx.Class.define('cv.ui.manager.model.Schema', {
  extend: qx.core.Object,

  /*
  ***********************************************
    CONSTRUCTOR
  ***********************************************
  */
  construct(filename) {
    super();
    if (!filename || !filename.match(/\.xsd$/)) {
      throw new Error('no, empty or invalid filename given, can not instantiate without one');
    }
    this.__filename = filename;
    this.setStructure(filename.endsWith('visu_config_tile.xsd') ? 'tile' : 'pure');

    this.__allowedRootElements = {};
    this.__referencedNodeCache = {};
    this.__typeNodeCache = {};
    this._cacheXSD();
  },

  /*
  ***********************************************
    STATICS
  ***********************************************
  */
  statics: {
    __CACHE: {},

    getInstance(schemaFile) {
      if (!Object.prototype.hasOwnProperty.call(this.__CACHE, schemaFile)) {
        this.__CACHE[schemaFile] = new cv.ui.manager.model.Schema(
          qx.util.ResourceManager.getInstance().toUri(schemaFile)
        );
      }
      return this.__CACHE[schemaFile];
    }
  },

  /*
  ***********************************************
    PROPERTIES
  ***********************************************
  */
  properties: {
    loaded: {
      check: 'Boolean',
      init: false,
      event: 'changeLoaded'
    },

    structure: {
      check: ['pure', 'tile'],
      apply: '_applyStructure'
    }
  },

  /*
  ***********************************************
    MEMBERS
  ***********************************************
  */
  members: {
    __filename: null,
    /**
     * object of the schema/xsd
     * @var object
     */
    __xsd: null,

    /**
     * object of allowed root-level elements
     * @var object
     */
    __allowedRootElements: null,

    /**
     * cache for referenced nods
     * @var object
     */
    __referencedNodeCache: null,

    /**
     * cache for getTypeNode
     * @var object
     */
    __typeNodeCache: null,

    /**
     * cache for #text-SchemaElement
     * @var object
     */
    __textNodeSchemaElement: null,

    /**
     * cache for #comment-SchemaElement
     * @var object
     */
    __commentNodeSchemaElement: null,

    /**
     * @var {Array<String>}
     */
    _widgetNames: null,

    /**
     * @var {String}
     */
    __rootElementName: null,

    /**
     * @var {String}
     */
    __pageParentElementName: null,

    /**
     * @var {String}
     */
    __pageElementName: null,

    _applyStructure(structure) {
      if (structure === 'tile') {
        this.__rootElementName = 'config';
        this.__pageParentElementName = 'main';
        this.__pageElementName = 'cv-page';
      } else {
        this.__rootElementName = 'pages';
        this.__pageElementName = 'page';
      }
    },

    onLoaded(callback, context) {
      if (this.isLoaded()) {
        callback.call(context);
      } else {
        this.addListenerOnce('changeLoaded', callback, context);
      }
    },

    /**
     * load and cache the xsd from the server
     */
    _cacheXSD() {
      const ajaxRequest = new qx.io.request.Xhr(this.__filename);
      ajaxRequest.set({
        accept: 'application/xml'
      });

      ajaxRequest.addListenerOnce('success', async (e) => {
        const req = e.getTarget();
        // Response parsed according to the server's response content type
        let xml = req.getResponse();
        if (xml && typeof xml === 'string') {
          xml = qx.xml.Document.fromString(xml);
        }
        // embed all includes
        let includeXml;
        for (let include of xml.querySelectorAll('schema include')) {
          includeXml = await this.loadXml('resource/' + include.getAttribute('schemaLocation'));
          const target = include.parentElement;
          include.remove();
          for (let includedChild of includeXml.querySelectorAll('schema > *')) {
            target.appendChild(includedChild);
          }
        }
        this.__xsd = xml;

        // parse the data, to have at least a list of root-level-elements
        this._parseXSD();
      });
      ajaxRequest.send();
    },

    async loadXml(file) {
      console.log('loading', file);
      return new Promise((resolve, reject) => {
        const ajaxRequest = new qx.io.request.Xhr(file);
        ajaxRequest.set({
          accept: 'application/xml'
        });
        ajaxRequest.addListenerOnce('success', e => {
          const req = e.getTarget();
          // Response parsed according to the server's response content type
          let xml = req.getResponse();
          if (xml && typeof xml === 'string') {
            xml = qx.xml.Document.fromString(xml);
          }
          resolve(xml);
        });
        ajaxRequest.addListenerOnce('statusError', e => {
          const req = e.getTarget();
          reject(new Error(req.getStatusText()));
        })
        ajaxRequest.send();
      });
    },

    /**
     * parse the schema once
     */
    _parseXSD() {
      // make a list of root-level elements
      this.__xsd.querySelectorAll('schema > element').forEach(element => {
        const name = element.getAttribute('name');
        this.__allowedRootElements[name] = new cv.ui.manager.model.schema.Element(element, this);
      });
      this.setLoaded(true);
    },

    getElementNode(name) {
      if (Object.prototype.hasOwnProperty.call(this.__allowedRootElements, name)) {
        return this.__allowedRootElements[name];
      }
      throw new Error('schema/xsd appears to be invalid, element ' + name + ' not allowed on root level');
    },

    /**
     * dive into the schema and find the element that is being pulled in by a ref.
     * Do so recursively.
     * referenced nodes can be top-level-nodes only!
     *
     * @param   type       string  Type of the node (e.g. element, attributeGroup, ...)
     * @param   refName    string  Name as per the ref-attribute
     * @param   noFallback boolean Don't look up other types as fallback, if the requested type is not found
     * @return  object          jQuery-object of the ref'ed element
     */
    getReferencedNode(type, refName, noFallback) {
      if (
        Object.prototype.hasOwnProperty.call(this.__referencedNodeCache, type) &&
        Object.prototype.hasOwnProperty.call(this.__referencedNodeCache[type], refName)
      ) {
        return this.__referencedNodeCache[type][refName];
      }
      const fallbackType = type === 'simpleType' ? 'complexType' : 'simpleType';
      if (!noFallback) {
        if (
          Object.prototype.hasOwnProperty.call(this.__referencedNodeCache, fallbackType) &&
          Object.prototype.hasOwnProperty.call(this.__referencedNodeCache[fallbackType], refName)
        ) {
          return this.__referencedNodeCache[fallbackType][refName];
        }
      }

      const selector = 'schema > ' + type + '[name="' + refName + '"]';
      let ref = this.__xsd.querySelector(selector);
      if (!ref && !noFallback) {
        try {
          ref = this.getReferencedNode(fallbackType, refName, true);
        } catch (e) {}
      }
      if (!ref) {
        throw new Error('schema/xsd appears to be invalid, reference ' + type + ' "' + refName + '" can not be found');
      }

      if (ref.hasAttribute('ref')) {
        // do it recursively, if necessary
        ref = this.getReferencedNode(type, ref.getAttribute('ref'));
      }

      if (!Object.prototype.hasOwnProperty.call(this.__referencedNodeCache, type)) {
        this.__referencedNodeCache[type] = {};
      }

      // fill the cache
      this.__referencedNodeCache[type][refName] = ref;

      return ref;
    },

    /**
     * get the definition of a type, be it complex or simple
     *
     * @param   type    string  Type of type to find (either simple or complex)
     * @param   name    string  Name of the type to find
     */
    getTypeNode(type, name) {
      if (
        Object.prototype.hasOwnProperty.call(this.__typeNodeCache, type) &&
        Object.prototype.hasOwnProperty.call(this.__typeNodeCache[type], name)
      ) {
        return this.__typeNodeCache[type][name];
      }
      let typeNode = this.__xsd.querySelector(type + 'Type[name="' + name + '"]');

      if (!typeNode) {
        throw new Error('schema/xsd appears to be invalid, ' + type + 'Type "' + name + '" can not be found');
      }

      if (typeof this.__typeNodeCache[type] == 'undefined') {
        this.__typeNodeCache[type] = {};
      }

      // fill the cache
      this.__typeNodeCache[type][name] = typeNode;

      return typeNode;
    },

    /**
     * get a SchemaElement for a #text-node
     *
     * @return  object  SchemaElement for #text-node
     */
    getTextNodeSchemaElement() {
      if (this.__textNodeSchemaElement === null) {
        // text-content is always a simple string
        const tmpXML = this.__xsd.createElement('element');
        tmpXML.setAttribute('name', '#text');
        tmpXML.setAttribute('type', 'xsd:string');
        this.__textNodeSchemaElement = new cv.ui.manager.model.schema.Element(tmpXML, this);
      }

      return this.__textNodeSchemaElement;
    },

    /**
     * get a SchemaElement for a #comment-node
     *
     * @return  object  SchemaElement for #comment-node
     */
    getCommentNodeSchemaElement() {
      if (this.__commentNodeSchemaElement === null) {
        // text-content is always a simple string
        const tmpXML = this.__xsd.createElement('element');
        tmpXML.setAttribute('name', '#comment');
        tmpXML.setAttribute('type', 'xsd:string');
        tmpXML.setAttribute('minOccurs', '0');
        tmpXML.setAttribute('maxOccurs', 'unbounded');
        this.__commentNodeSchemaElement = new cv.ui.manager.model.schema.Element(tmpXML, this);
      }

      return this.__commentNodeSchemaElement;
    },

    /**
     * get the DOM for this Schema
     *
     * @return  object  DOM
     */
    getSchemaDOM() {
      return this.__xsd;
    },

    /**
     * A CometVisu-Schema specific helper function that returns an array of all widget names.
     * @returns {Array<String>}
     */
    getWidgetNames() {
      if (!this._widgetNames) {
        const root = this.getElementNode(this.__rootElementName);
        let pageParent = root;
        if (this.__pageParentElementName) {
          pageParent = root.getSchemaElementForElementName(this.__pageParentElementName);
        }
        const page = pageParent.getSchemaElementForElementName(this.__pageElementName);

        this._widgetNames = Object.keys(page.getAllowedElements()).filter(
          name => !name.startsWith('#') && name !== 'layout'
        );
      }
      return this._widgetNames;
    },

    isRoot(name) {
      return name === this.__rootElementName;
    },

    isPage(name) {
      return (name = this.__pageElementName);
    }
  },

  /*
  ***********************************************
    DESTRUCTOR
  ***********************************************
  */
  destruct() {
    this.__xsd = null;
    this._disposeObjects('__commentNodeSchemaElement', '__textNodeSchemaElement');

    this._disposeMap('__allowedRootElements');
    this.__referencedNodeCache = null;
    this.__typeNodeCache = null;
    this._widgetNames = null;
  }
});

File: utils/docutils/directives/helper/schema.py
# -*- coding: utf-8 -*-

# copyright (c) 2010-2016, Christian Mayer and the CometVisu contributers.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the Free
# Software Foundation; either version 3 of the License, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA

from lxml.etree import parse
import re
SCHEMA_SPACE = "{http://www.w3.org/2001/XMLSchema}"
XML_SPACE = "{http://www.w3.org/XML/1998/namespace}"


class Schema:
    """
    Parse visu_config.xsd and retrieve relevant information for a widget

    @author Tobias Bräutigam
    @since 0.10.0
    """

    def __init__(self, schemafile):
        self.root = parse(schemafile)

    def findall(self, path):
        return self.root.findall( path.replace("xs:", SCHEMA_SPACE) )

    def find(self, path):
        return self.root.find( path.replace("xs:", SCHEMA_SPACE) )

    def names_of(self, nodes):
        return [node.get("name") for node in nodes]

    def get_Types(self, t_name):
        return self.names_of( self.findall(t_name) )

    def get_simpleTypes(self):
        return self.get_Types("xs:simpleType")

    def get_complexTypes(self):
        return self.get_Types("xs:complexType")

    def get_widget(self, widget_name):
        return self.findall("xs:complexType[@name='%s']" % widget_name)

    def get_widget_attributes(self, widget_name, type_node=None):
        res = self.findall("xs:complexType[@name='%s']/xs:attribute" % widget_name)
        res += self.findall("xs:complexType[@name='%s']/xs:simpleContent/xs:extension/xs:attribute" % widget_name)
        if len(res) == 0 and type_node is not None:
            res = type_node.findall("xs:complexType/xs:attribute".replace("xs:", SCHEMA_SPACE))
            res += type_node.findall("xs:complexType/xs:simpleContent/xs:extension/xs:attribute".replace("xs:", SCHEMA_SPACE))
            ext = type_node.find(".//xs:simpleContent/xs:extension".replace("xs:", SCHEMA_SPACE))
            if ext is not None and ext.get("base") is not None:
                # get all attributes from extension
                for attr in self.get_widget_attributes(ext.get("base")):
                    res.append(attr)

        # find and resolve attribute groups
        res += self.get_attribute_groups("xs:complexType[@name='%s']/xs:attributeGroup" % widget_name)
        res += self.get_attribute_groups("xs:complexType[@name='%s']/xs:simpleContent/xs:extension/xs:attributeGroup" % widget_name)

        # find and resolve inherited attributes
        ext = self.find("xs:complexType[@name='%s']/xs:complexContent/xs:extension" % widget_name)
        if ext is not None:
            # get attributes from extended element
            res += self.get_widget_attributes(ext.get("base"))
            # get attributes from extension
            res += ext.findall(".//xs:attribute".replace("xs:", SCHEMA_SPACE))

        return res

    def get_attribute_groups(self, query):
        attrs = []
        for group in self.findall(query):
            if "ref" in group.attrib:
                name = group.get('ref')
                attrs += self.get_group_attributes(name)
        return attrs

    def get_attribute(self, widget_name):
        return self.findall("xs:attribute[@name='%s']" % widget_name)[0]

    def get_group_attributes(self, widget_name):
        return self.findall("xs:attributeGroup[@name='%s']/xs:attribute" % widget_name)

    def get_widget_elements(self, widget_name, locale='en', ctype=None, sub_type=False):
        if ctype is None:
            ctype = self.find("xs:complexType[@name='%s']" % widget_name)
        if ctype is None:
            return []
        elems = ctype.findall(".//xs:element".replace("xs:", SCHEMA_SPACE))
        simple_ext = ctype.find("./xs:simpleContent/xs:extension".replace("xs:", SCHEMA_SPACE))
        complex_ext = ctype.find("./xs:complexContent/xs:extension".replace("xs:", SCHEMA_SPACE))
        elem_names = [x.get('name') for x in elems]

        # find extended elements
        for elem in elems:
            if not isinstance(elem, tuple) and elem.get('type') is None:
                sub_type = elem.find("./xs:complexType".replace("xs:", SCHEMA_SPACE))
                if sub_type is not None:
                    for sub_elem in self.get_widget_elements(elem.get('name'), ctype=sub_type, sub_type=True):
                         name = elem.get('name') if not isinstance(sub_elem, tuple) else sub_elem[0]
                         if name not in elem_names:
                             elems.append(sub_elem)
                             elem_names.append(name)

        if simple_ext is not None and simple_ext.get("base") is not None:
            # should we really hardcode this?
            ref = self.find("xs:simpleType[@name='%s']" % simple_ext.get("base"))
            doc = self.get_node_documentation(ref, locale).text if ref is not None and self.get_node_documentation(ref, locale) is not None else ""
            elems.append(("#text", "string", doc))
        elif ctype.get("mixed", "false") == "true":
            doc = self.get_node_documentation(ctype, locale).text if self.get_node_documentation(ctype, locale) is not None else ""
            elems.append(("#text", "string", doc))

        if complex_ext is not None and complex_ext.get("base") is not None:
            for e in self.get_widget_elements(complex_ext.get("base"), locale):
                name = e.get('name') if not isinstance(e, tuple) else e[0]
                if name not in elem_names:
                    elems.append(e)
                    elem_names.append(name)

        return elems

    def get_elements_of_attribute(self, attribute):
        return self.names_of(self.findall(".//xs:element/xs:complexType/xs:" + attribute + "/../.."))

    def get_node_documentation(self, node, locale):
        return node.find("%sannotation/%sdocumentation[@%slang='%s']" % (SCHEMA_SPACE, SCHEMA_SPACE, XML_SPACE, locale))

    def get_attribute_type(self, node, locale):
        a_type = None
        values = []
        enums = None
        description = self.get_node_documentation(node, locale)
        if 'type' in node.attrib:
            a_type = node.get('type')
            if a_type[0:4] != "xsd:":
                # check if this is a type defined in the document
                external_type = self.find("xs:simpleType[@name='"+a_type+"']".replace("xs:", SCHEMA_SPACE))
                if external_type is not None:
                    enums = external_type.findall("xs:restriction/xs:enumeration".replace("xs:", SCHEMA_SPACE))
                    values = [enum.get('value') for enum in enums]
                    a_type = external_type.find("xs:restriction".replace("xs:", SCHEMA_SPACE)).get("base")
                    if description is None:
                        description = self.get_node_documentation(external_type, locale)
        elif len(node.findall("xs:simpleType/xs:restriction".replace("xs:", SCHEMA_SPACE))) > 0:
            enums = node.findall("xs:simpleType/xs:restriction/xs:enumeration".replace("xs:", SCHEMA_SPACE))
            values = [enum.get('value') for enum in enums]
            a_type = node.find("xs:simpleType/xs:restriction".replace("xs:", SCHEMA_SPACE)).get("base")
        if description is not None:
            description = re.sub(r"\n\s+", " ", description.text).strip()
            return a_type, values, enums, description
        return a_type, values, enums, ''

    def get_element_attributes(self, name):

        node = self.find(".//xs:element[@name='" + name + "']")
        if node is None:
            node = self.find(".//xs:complexType[@name='" + name + "']")
        if node is None:
            return None
        else:
            return node.attrib
